File: examples/python/metaquest.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import signal
import sys
import threading
import time
from dataclasses import dataclass, field

import igris_c_sdk as igc_sdk

DEFAULT_DOMAIN_ID = 0
DEFAULT_TOPIC = "igris_c/metaquest"
FINGER_NAMES = ("thumb", "index", "middle", "ring", "pinky")


def format_array(values: tuple[float, ...], precision: int = 3) -> str:
    return "[" + ", ".join(f"{value:.{precision}f}" for value in values) + "]"


def format_bool(value: bool) -> str:
    return "Y" if value else "N"


def format_bool_tuple(values: tuple[bool, ...]) -> str:
    return "[" + ", ".join(format_bool(value) for value in values) + "]"


def uses_controller_tracking(tracking_mode: int) -> bool:
    return tracking_mode == 0


def uses_hand_tracking(tracking_mode: int) -> bool:
    return tracking_mode == 1


def print_rule(width: int = 108) -> None:
    sys.stdout.write("=" * width + "\n")


def print_kv(key: str, value: str) -> None:
    sys.stdout.write(f"{key:<20} | {value}\n")


def vec3_tuple(vec: object) -> tuple[float, float, float]:
    data = vec.data()
    return (float(data[0]), float(data[1]), float(data[2]))


def vec4_tuple(vec: object) -> tuple[float, float, float, float]:
    data = vec.data()
    return (float(data[0]), float(data[1]), float(data[2]), float(data[3]))


def float_tuple(values: object) -> tuple[float, ...]:
    return tuple(float(value) for value in values)


def bool_tuple(values: object) -> tuple[bool, ...]:
    return tuple(bool(value) for value in values)


def as_sequence(value: object) -> list[object]:
    try:
        return list(value)
    except TypeError:
        return [value]


@dataclass
class HeaderSnapshot:
    seq: int
    sec: int
    nanosec: int
    frame_id: str


@dataclass
class ControllerSnapshot:
    position: tuple[float, float, float]
    quaternion: tuple[float, float, float, float]
    euler: tuple[float, float, float]
    linear_velocity: tuple[float, float, float]
    angular_velocity: tuple[float, float, float]
    joystick: tuple[float, float]
    touchpad: tuple[float, float]
    index_trigger: float
    hand_trigger: float
    button_one: bool
    button_two: bool
    button_three: bool
    button_four: bool
    joystick_click: bool
    index_trigger_touch: bool
    hand_trigger_touch: bool

    def buttons_text(self) -> str:
        return (
            f"1:{format_bool(self.button_one)} "
            f"2:{format_bool(self.button_two)} "
            f"3:{format_bool(self.button_three)} "
            f"4:{format_bool(self.button_four)} "
            f"JC:{format_bool(self.joystick_click)} "
            f"IT:{format_bool(self.index_trigger_touch)} "
            f"HT:{format_bool(self.hand_trigger_touch)}"
        )


@dataclass
class FingerSnapshot:
    name: str
    pinch: bool
    pinch_strength: float
    finger_confidence: float
    tip_valid: bool
    tip_position: tuple[float, float, float]
    tip_quaternion: tuple[float, float, float, float]
    tip_to_thumb_distance: float


@dataclass
class HandSnapshot:
    position: tuple[float, float, float]
    quaternion: tuple[float, float, float, float]
    euler: tuple[float, float, float]
    confidence: float
    pinch: tuple[bool, bool, bool, bool, bool]
    tip_valid: tuple[bool, bool, bool, bool, bool]
    fingers: list[FingerSnapshot]


@dataclass
class QuestSnapshot:
    header: HeaderSnapshot
    tracking_mode: int
    head_position: tuple[float, float, float]
    head_quaternion: tuple[float, float, float, float]
    head_euler: tuple[float, float, float]
    controllers: list[ControllerSnapshot]
    hands: list[HandSnapshot]
    ip: str
    battery: float


def snapshot_from_message(message: object) -> QuestSnapshot:
    header = message.header()
    controllers = [
        ControllerSnapshot(
            position=vec3_tuple(controller.position()),
            quaternion=vec4_tuple(controller.quaternion()),
            euler=vec3_tuple(controller.euler()),
            linear_velocity=float_tuple(controller.linear_velocity()),
            angular_velocity=float_tuple(controller.angular_velocity()),
            joystick=float_tuple(controller.joystick()),
            touchpad=float_tuple(controller.touchpad()),
            index_trigger=float(controller.index_trigger()),
            hand_trigger=float(controller.hand_trigger()),
            button_one=bool(controller.button_one()),
            button_two=bool(controller.button_two()),
            button_three=bool(controller.button_three()),
            button_four=bool(controller.button_four()),
            joystick_click=bool(controller.joystick_click()),
            index_trigger_touch=bool(controller.index_trigger_touch()),
            hand_trigger_touch=bool(controller.hand_trigger_touch()),
        )
        for controller in as_sequence(message.controller())
    ]
    hands: list[HandSnapshot] = []
    for hand in as_sequence(message.hand()):
        tip_positions = hand.tip_position()
        tip_quaternions = hand.tip_quaternion()
        fingers = [
            FingerSnapshot(
                name=FINGER_NAMES[index],
                pinch=bool(hand.pinch()[index]),
                pinch_strength=float(hand.pinch_strength()[index]),
                finger_confidence=float(hand.finger_confidence()[index]),
                tip_valid=bool(hand.tip_valid()[index]),
                tip_position=vec3_tuple(tip_positions[index]),
                tip_quaternion=vec4_tuple(tip_quaternions[index]),
                tip_to_thumb_distance=float(hand.tip_to_thumb_distance()[index]),
            )
            for index in range(5)
        ]
        hands.append(
            HandSnapshot(
                position=vec3_tuple(hand.position()),
                quaternion=vec4_tuple(hand.quaternion()),
                euler=vec3_tuple(hand.euler()),
                confidence=float(hand.confidence()),
                pinch=bool_tuple(hand.pinch()),
                tip_valid=bool_tuple(hand.tip_valid()),
                fingers=fingers,
            )
        )

    return QuestSnapshot(
        header=HeaderSnapshot(
            seq=int(header.seq()),
            sec=int(header.sec()),
            nanosec=int(header.nanosec()),
            frame_id=str(header.frame_id()),
        ),
        tracking_mode=int(message.tracking_mode()),
        head_position=vec3_tuple(message.head_position()),
        head_quaternion=vec4_tuple(message.head_quaternion()),
        head_euler=vec3_tuple(message.head_euler()),
        controllers=controllers,
        hands=hands,
        ip=str(message.ip()),
        battery=float(message.battery()),
    )


@dataclass
class SubscriberState:
    topic: str
    lock: threading.Lock = field(default_factory=threading.Lock)
    message_count: int = 0
    last_receive: float = 0.0
    snapshot: QuestSnapshot | None = None

    def on_message(self, message: object) -> None:
        now = time.perf_counter()
        snapshot = snapshot_from_message(message)

        with self.lock:
            self.last_receive = now
            self.message_count += 1
            self.snapshot = snapshot

    def view(self) -> tuple[int, float, float, QuestSnapshot | None]:
        with self.lock:
            return self.message_count, self.last_receive, self.snapshot


def print_snapshot(domain_id: int, state: SubscriberState) -> None:
    message_count, last_receive, snapshot = state.view()
    sys.stdout.write("\x1b[2J\x1b[H")
    print_rule()
    sys.stdout.write("MetaQuest Raw Viewer\n")
    print_rule()
    print_kv("Domain", str(domain_id))
    print_kv("Topic", state.topic)

    if snapshot is None:
        print_kv("Status", "Waiting for QuestMessage samples...")
        print_kv("Quit", "Ctrl+C")
        print_rule()
        sys.stdout.flush()
        return

    age_ms = int((time.perf_counter() - last_receive) * 1000.0) if last_receive > 0.0 else 0
    print_kv("Message Count", str(message_count))
    print_kv("Age", f"{age_ms} ms")
    print_kv("Seq", str(snapshot.header.seq))
    print_kv("Stamp", f"{snapshot.header.sec}.{snapshot.header.nanosec:09d}")
    print_kv("Frame ID", snapshot.header.frame_id)
    print_kv("IP", snapshot.ip)
    print_kv("Battery", f"{snapshot.battery:.1f}")
    print_kv("Tracking Mode", str(snapshot.tracking_mode))
    if uses_controller_tracking(snapshot.tracking_mode):
        print_kv("Active Input", "controller")
    elif uses_hand_tracking(snapshot.tracking_mode):
        print_kv("Active Input", "hand")
    else:
        print_kv("Active Input", "unknown")
    print_rule()
    sys.stdout.write("Pose Summary\n")
    print_rule()
    print_kv("Head Pos", format_array(snapshot.head_position))
    print_kv("Head Euler", format_array(snapshot.head_euler))
    print_kv("Head Quat", format_array(snapshot.head_quaternion))
    if uses_controller_tracking(snapshot.tracking_mode):
        for index, controller in enumerate(snapshot.controllers):
            prefix = f"Ctrl[{index}]"
            print_kv(f"{prefix} Pos", format_array(controller.position))
            print_kv(f"{prefix} Euler", format_array(controller.euler))
            print_kv(f"{prefix} Quat", format_array(controller.quaternion))
            print_kv(f"{prefix} Lin Vel", format_array(controller.linear_velocity))
            print_kv(f"{prefix} Ang Vel", format_array(controller.angular_velocity))
            print_kv(f"{prefix} Joystick", format_array(controller.joystick))
            print_kv(f"{prefix} Touchpad", format_array(controller.touchpad))
            print_kv(f"{prefix} Index Trg", f"{controller.index_trigger:.3f}")
            print_kv(f"{prefix} Hand Trg", f"{controller.hand_trigger:.3f}")
            print_kv(f"{prefix} Buttons", controller.buttons_text())
    elif uses_hand_tracking(snapshot.tracking_mode):
        for index, hand in enumerate(snapshot.hands):
            prefix = f"Hand[{index}]"
            print_kv(f"{prefix} Pos", format_array(hand.position))
            print_kv(f"{prefix} Euler", format_array(hand.euler))
            print_kv(f"{prefix} Quat", format_array(hand.quaternion))
            print_kv(f"{prefix} Conf", f"{hand.confidence:.3f}")
            print_kv(f"{prefix} Pinch", format_bool_tuple(hand.pinch))
            print_kv(f"{prefix} Tip Valid", format_bool_tuple(hand.tip_valid))
            print_kv(f"{prefix} Fingers", str(list(FINGER_NAMES)))
            print_rule()
            sys.stdout.write(f"{prefix} Finger Table\n")
            print_rule()
            sys.stdout.write(
                f"{'Finger':<8} | {'Pinch':<5} | {'Strength':<8} | {'Conf':<8} | "
                f"{'Valid':<8} | {'Dist':<10} | {'Tip Pos':<22} | Tip Quat\n"
            )
            print_rule()
            for finger in hand.fingers:
                sys.stdout.write(
                    f"{finger.name:<8} | {format_bool(finger.pinch):<5} | {finger.pinch_strength:<8.3f} | "
                    f"{finger.finger_confidence:<8.3f} | {format_bool(finger.tip_valid):<8} | "
                    f"{finger.tip_to_thumb_distance:<10.3f} | {format_array(finger.tip_position):<22} | "
                    f"{format_array(finger.tip_quaternion)}\n"
                )
            print_rule()
    else:
        print_kv("Tracking Data", "Unsupported tracking_mode")
    print_kv("Quit", "Ctrl+C")
    print_rule()
    sys.stdout.flush()


def build_arg_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="MetaQuest raw QuestMessage viewer")
    parser.add_argument(
        "--domain-id",
        type=int,
        default=DEFAULT_DOMAIN_ID,
        help=f"Cyclone DDS domain id (default: {DEFAULT_DOMAIN_ID})",
    )
    parser.add_argument(
        "topic",
        nargs="?",
        default=DEFAULT_TOPIC,
        help=f"QuestMessage topic name (default: {DEFAULT_TOPIC})",
    )
    return parser


def main() -> int:
    args = build_arg_parser().parse_args()
    state = SubscriberState(topic=args.topic)
    stop_event = threading.Event()

    def handle_signal(_signum: int, _frame: object) -> None:
        stop_event.set()

    signal.signal(signal.SIGINT, handle_signal)
    signal.signal(signal.SIGTERM, handle_signal)

    channel = igc_sdk.ChannelFactory.Instance()
    channel.Init(args.domain_id)

    subscriber = igc_sdk.QuestMessageSubscriber(args.topic)
    if not subscriber.init(state.on_message):
        print("Failed to initialize QuestMessageSubscriber", file=sys.stderr)
        channel.Release()
        return 1

    try:
        while not stop_event.is_set():
            print_snapshot(args.domain_id, state)
            time.sleep(0.1)
    finally:
        subscriber.stop()
        channel.Release()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/metaquest.cpp
#include "igris_sdk/igris_c_msgs.hpp"
#include "igris_sdk/types.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdlib>
#include <dds/dds.hpp>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>

namespace {

using QuestMessage = igris_sdk::QuestMessage;

constexpr int kDefaultDomainId      = 0;
constexpr const char *kDefaultTopic = "igris_c/metaquest";

std::atomic<bool> g_stop{false};

void handle_signal(int) { g_stop.store(true); }

template <size_t N> std::string format_array(const std::array<float, N> &values, int precision = 3) {
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(precision) << "[";
    for (size_t i = 0; i < N; ++i) {
        if (i > 0) {
            oss << ", ";
        }
        oss << values[i];
    }
    oss << "]";
    return oss.str();
}

std::string format_bool(bool value) { return value ? "Y" : "N"; }

template <size_t N> std::string format_bool_array(const std::array<bool, N> &values) {
    std::ostringstream oss;
    oss << "[";
    for (size_t i = 0; i < N; ++i) {
        if (i > 0) {
            oss << ", ";
        }
        oss << format_bool(values[i]);
    }
    oss << "]";
    return oss.str();
}

bool uses_controller_tracking(int tracking_mode) { return tracking_mode == 0; }

bool uses_hand_tracking(int tracking_mode) { return tracking_mode == 1; }

void print_rule(size_t width);
void print_kv(const std::string &key, const std::string &value);
void print_usage(const char *program);

std::string format_buttons(const igris_sdk::QuestController &controller) {
    std::ostringstream oss;
    oss << "1:" << format_bool(controller.button_one()) << " " << "2:" << format_bool(controller.button_two()) << " "
        << "3:" << format_bool(controller.button_three()) << " " << "4:" << format_bool(controller.button_four()) << " "
        << "JC:" << format_bool(controller.joystick_click()) << " " << "IT:" << format_bool(controller.index_trigger_touch()) << " "
        << "HT:" << format_bool(controller.hand_trigger_touch());
    return oss.str();
}

std::string format_finger_names() { return "[thumb, index, middle, ring, pinky]"; }

void print_controller_section(size_t index, const igris_sdk::QuestController &controller) {
    const std::string prefix = "Ctrl[" + std::to_string(index) + "]";
    print_kv(prefix + " Pos", format_array(controller.position().data()));
    print_kv(prefix + " Euler", format_array(controller.euler().data()));
    print_kv(prefix + " Quat", format_array(controller.quaternion().data()));
    print_kv(prefix + " Lin Vel", format_array(controller.linear_velocity()));
    print_kv(prefix + " Ang Vel", format_array(controller.angular_velocity()));
    print_kv(prefix + " Joystick", format_array(controller.joystick()));
    print_kv(prefix + " Touchpad", format_array(controller.touchpad()));
    {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(3) << controller.index_trigger();
        print_kv(prefix + " Index Trg", oss.str());
    }
    {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(3) << controller.hand_trigger();
        print_kv(prefix + " Hand Trg", oss.str());
    }
    print_kv(prefix + " Buttons", format_buttons(controller));
}

void print_hand_section(size_t index, const igris_sdk::QuestHand &hand) {
    const std::string prefix = "Hand[" + std::to_string(index) + "]";
    print_kv(prefix + " Pos", format_array(hand.position().data()));
    print_kv(prefix + " Euler", format_array(hand.euler().data()));
    print_kv(prefix + " Quat", format_array(hand.quaternion().data()));
    {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(3) << hand.confidence();
        print_kv(prefix + " Conf", oss.str());
    }
    print_kv(prefix + " Pinch", format_bool_array(hand.pinch()));
    print_kv(prefix + " Tip Valid", format_bool_array(hand.tip_valid()));
    print_kv(prefix + " Fingers", format_finger_names());
    print_rule(108);
    std::cout << prefix << " Finger Table\n";
    print_rule(108);
    std::cout << std::left << std::setw(8) << "Finger" << " | " << std::setw(5) << "Pinch" << " | " << std::setw(8) << "Strength" << " | "
              << std::setw(8) << "Conf" << " | " << std::setw(8) << "Valid" << " | " << std::setw(10) << "Dist" << " | " << std::setw(22)
              << "Tip Pos" << " | " << "Tip Quat" << "\n";
    print_rule(108);

    static constexpr const char *finger_names[5] = {"thumb", "index", "middle", "ring", "pinky"};
    for (size_t i = 0; i < 5; ++i) {
        std::ostringstream strength;
        std::ostringstream confidence;
        std::ostringstream distance;
        strength << std::fixed << std::setprecision(3) << hand.pinch_strength()[i];
        confidence << std::fixed << std::setprecision(3) << hand.finger_confidence()[i];
        distance << std::fixed << std::setprecision(3) << hand.tip_to_thumb_distance()[i];

        std::cout << std::left << std::setw(8) << finger_names[i] << " | " << std::setw(5) << format_bool(hand.pinch()[i]) << " | "
                  << std::setw(8) << strength.str() << " | " << std::setw(8) << confidence.str() << " | " << std::setw(8)
                  << format_bool(hand.tip_valid()[i]) << " | " << std::setw(10) << distance.str() << " | " << std::setw(22)
                  << format_array(hand.tip_position()[i].data()) << " | " << format_array(hand.tip_quaternion()[i].data()) << "\n";
    }
}

void print_controller_payload(const igris_sdk::QuestController &controller) { print_controller_section(0, controller); }

template <size_t N> void print_controller_payload(const std::array<igris_sdk::QuestController, N> &controllers) {
    for (size_t i = 0; i < controllers.size(); ++i) {
        print_controller_section(i, controllers[i]);
        if (i + 1 < controllers.size()) {
            print_rule(108);
        }
    }
}

void print_hand_payload(const igris_sdk::QuestHand &hand) {
    print_hand_section(0, hand);
    print_rule(108);
}

template <size_t N> void print_hand_payload(const std::array<igris_sdk::QuestHand, N> &hands) {
    for (size_t i = 0; i < hands.size(); ++i) {
        print_hand_section(i, hands[i]);
        if (i + 1 < hands.size()) {
            print_rule(108);
        }
    }
    print_rule(108);
}

struct Snapshot {
    bool has_message = false;
    QuestMessage message;
    uint64_t message_count                             = 0;
    std::chrono::steady_clock::time_point last_receive = std::chrono::steady_clock::now();
};

void print_rule(size_t width) { std::cout << std::string(width, '=') << "\n"; }

void print_kv(const std::string &key, const std::string &value) {
    std::cout << std::left << std::setw(20) << key << " | " << value << "\n";
}

void print_snapshot(int domain_id, const std::string &topic, const Snapshot &snapshot) {
    std::cout << "\x1B[2J\x1B[H";
    print_rule(108);
    std::cout << "MetaQuest Raw Viewer\n";
    print_rule(108);
    print_kv("Domain", std::to_string(domain_id));
    print_kv("Topic", topic);

    if (!snapshot.has_message) {
        print_kv("Status", "Waiting for QuestMessage samples...");
        print_kv("Quit", "Ctrl+C");
        print_rule(108);
        std::cout << std::flush;
        return;
    }

    const auto age_ms =
        std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - snapshot.last_receive).count();
    const auto &message = snapshot.message;
    const auto &header  = message.header();

    {
        std::ostringstream oss;
        oss << snapshot.message_count;
        print_kv("Message Count", oss.str());
    }
    print_kv("Age", std::to_string(age_ms) + " ms");
    print_kv("Seq", std::to_string(header.seq()));
    {
        std::ostringstream oss;
        oss << header.sec() << "." << std::setw(9) << std::setfill('0') << header.nanosec() << std::setfill(' ');
        print_kv("Stamp", oss.str());
    }
    print_kv("Frame ID", header.frame_id());
    print_kv("IP", message.ip());
    {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(1) << message.battery();
        print_kv("Battery", oss.str());
    }
    print_kv("Tracking Mode", std::to_string(message.tracking_mode()));
    if (uses_controller_tracking(message.tracking_mode())) {
        print_kv("Active Input", "controller");
    } else if (uses_hand_tracking(message.tracking_mode())) {
        print_kv("Active Input", "hand");
    } else {
        print_kv("Active Input", "unknown");
    }
    print_rule(108);
    std::cout << "Pose Summary\n";
    print_rule(108);
    print_kv("Head Pos", format_array(message.head_position().data()));
    print_kv("Head Euler", format_array(message.head_euler().data()));
    print_kv("Head Quat", format_array(message.head_quaternion().data()));
    if (uses_controller_tracking(message.tracking_mode())) {
        print_controller_payload(message.controller());
    } else if (uses_hand_tracking(message.tracking_mode())) {
        print_hand_payload(message.hand());
    } else {
        print_kv("Tracking Data", "Unsupported tracking_mode");
    }
    print_rule(108);
    print_kv("Quit", "Ctrl+C");
    print_rule(108);
    std::cout << std::flush;
}

bool parse_domain_id(const char *value, int &domain_id) {
    char *end   = nullptr;
    long parsed = std::strtol(value, &end, 10);
    if (end == value || *end != '\0') {
        return false;
    }
    domain_id = static_cast<int>(parsed);
    return true;
}

bool parse_arguments(int argc, char **argv, int &domain_id, std::string &topic) {
    bool domain_id_set = false;
    bool topic_set     = false;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];

        if (arg == "-h" || arg == "--help") {
            print_usage(argv[0]);
            std::exit(0);
        }

        if (arg == "--domain-id") {
            if (i + 1 >= argc || !parse_domain_id(argv[++i], domain_id)) {
                std::cerr << "Invalid --domain-id value\n";
                return false;
            }
            domain_id_set = true;
            continue;
        }

        if (arg == "--topic") {
            if (i + 1 >= argc) {
                std::cerr << "Missing value for --topic\n";
                return false;
            }
            topic     = argv[++i];
            topic_set = true;
            continue;
        }

        if (!domain_id_set && parse_domain_id(arg.c_str(), domain_id)) {
            domain_id_set = true;
            continue;
        }

        if (!topic_set) {
            topic     = arg;
            topic_set = true;
            continue;
        }

        std::cerr << "Unexpected argument: " << arg << "\n";
        return false;
    }

    return true;
}

void print_usage(const char *program) {
    std::cout << "Usage: " << program << " [--domain-id N] [--topic TOPIC]\n"
              << "       " << program << " [domain_id] [topic]\n"
              << "Defaults: domain_id=" << kDefaultDomainId << ", topic=" << kDefaultTopic << "\n";
}

}  // namespace

int main(int argc, char **argv) {
    std::signal(SIGINT, handle_signal);
    std::signal(SIGTERM, handle_signal);

    int domain_id     = kDefaultDomainId;
    std::string topic = kDefaultTopic;

    if (!parse_arguments(argc, argv, domain_id, topic)) {
        print_usage(argv[0]);
        return 1;
    }

    try {
        dds::domain::DomainParticipant participant(domain_id);
        dds::topic::Topic<QuestMessage> dds_topic(participant, topic);
        dds::sub::Subscriber subscriber(participant);

        dds::sub::qos::DataReaderQos reader_qos = subscriber.default_datareader_qos();
        reader_qos << dds::core::policy::Reliability::Reliable() << dds::core::policy::History::KeepLast(10);
        dds::sub::DataReader<QuestMessage> reader(subscriber, dds_topic, reader_qos);

        Snapshot snapshot;
        while (!g_stop.load()) {
            auto samples = reader.take();
            for (const auto &sample : samples) {
                if (!sample.info().valid()) {
                    continue;
                }

                snapshot.message = sample.data();
                snapshot.message_count++;
                snapshot.last_receive = std::chrono::steady_clock::now();
                snapshot.has_message  = true;
            }

            print_snapshot(domain_id, topic, snapshot);
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    } catch (const dds::core::Exception &e) {
        std::cerr << "Failed to initialize DDS subscriber for topic " << topic << ": " << e.what() << "\n";
        return 1;
    }
    return 0;
}
